R/SandwichLayerVariance.R

```diff
@@ -558,3 +559,459 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
+  cov0l <- lapply(1:length(cov0), 
+                  function(s) if (is.na(cov0[[s]][1,1])) covll[[s]] else cov0[[s]])
+  
+  V00u <- Reduce('+', cov0u)
+  V00l <- Reduce('+', cov0l)
+  return(rbind(V00u, V00l))
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @keywords internal
+.add_mat_diag <- function(A, B){
+  d <- nrow(A)
+  A <- matrix(rep(diag(A), d), nrow = d, byrow = FALSE)
+  B <- matrix(rep(diag(B), d), nrow = d, byrow = TRUE)
+  #return(sqrt(abs(A * B)))
+  return((A + B) / 2)
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @keywords internal
+.add_vec <- function(a, upper = TRUE){
+  if (nrow(a) > 1) return(0)
+  a <- as.numeric(a)
+  d <- length(a)
+  A <- matrix(rep(a, d), nrow = d, byrow = FALSE)
+  B <- matrix(rep(a, d), nrow = d, byrow = TRUE)
+  #if (upper) return(A*B + sqrt(abs(A*B)))
+  #else return(A*B - sqrt(abs(A*B)))
+  if (upper) return((A + B)^2 / 2)
+  else return(- (A - B)^2 / 2)
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @keywords internal
+.cov01_est <- function(XX, zobs, bid){
+  cov0 <- tapply(1:nrow(XX[zobs==0,]), bid[zobs==0], function(s) cov(XX[zobs==0,][s,]))
+  cov1 <- tapply(1:nrow(XX[zobs==1,]), bid[zobs==1], function(s) cov(XX[zobs==1,][s,]))
+  cov01 <- lapply(1:length(cov0), function(s) .add_mat_diag(cov0[[s]], cov1[[s]]))
+  
+  cov01u <- lapply(1:length(cov0), 
+                   function(s) if (!is.na(cov01[[s]][1,1])) cov01[[s]]
+                   else .add_mat_sqdif(XX, zobs, bid, s))
+  cov01l <- lapply(1:length(cov0), 
+                   function(s) if (!is.na(cov01[[s]][1,1])) -cov01[[s]]
+                   else -.add_mat_sqdif(XX, zobs, bid, s, upper=FALSE))
+  V01u <- Reduce('+', cov01u)
+  V01l <- Reduce('+', cov01l)
+  return(rbind(V01u, V01l))
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @keywords internal
+.add_mat_sqdif <- function(X, zobs, bid, b, upper = TRUE){
+  A <- X[zobs==0 & bid==b, ]
+  if (upper) B <- X[zobs==1 & bid==b, ] else B <- -X[zobs==1 & bid==b, ]
+  cov01u <- matrix(0, nrow=ncol(A), ncol=ncol(A))
+  for (j in 1:ncol(B)){
+    for (h in 1:nrow(B)){
+      cov01u[,j] <- cov01u[,j] + colSums((A + B[h,j])^2)
+    }
+  }
+  cov01u <- cov01u / nrow(A) / nrow(B)
+  return(cov01u)
+}
+
+#' @title (Internal) Calculate grave{phi}
+#' @keywords internal
+#' @param x a fitted \code{teeMod} model
+.get_phi_tilde <- function(x, ...){
+  ws <- x$weights
+  n <- length(ws) # number of units in Q
+  design_obj <- x@Design
+  df <- x$call$data
+  
+  # treatment assignments
+  name_trt <- colnames(design_obj@structure)[design_obj@column_index == "t"]
+  assignment <- df[[name_trt]]
+  k <- length(unique(assignment))
+  z_ind <- sapply(c(0,1), function(val) as.integer(assignment == val))
+  
+  # stratum ids 
+  name_blk <- colnames(design_obj@structure)[design_obj@column_index == "b"]
```

Comment:
Same note about `var_names()` where possible

Reply:
Updated in commit https://github.com/benbhansen-stats/propertee/commit/6eb8927664674088f61b3323683e7d592a2de431